### ISSUE TITLE
LSTM Model Architecture Implementation - Bidirectional LSTM with Configurable Parameters

### ISSUE THREAD

**Issue by dhar174** [dhar174/steel_defect_demo#62]
# LSTM Model Architecture Implementation

## Overview
Implement a comprehensive Long Short-Term Memory (LSTM) neural network architecture for sequence-based steel casting defect prediction. This task focuses on creating a robust, configurable LSTM model that can effectively capture temporal patterns in sensor data for improved defect detection performance.

## Objectives
- Enhance the existing `SteelDefectLSTM` class with advanced architectural features
- Implement bidirectional LSTM layers with configurable hidden size and depth
- Add comprehensive regularization techniques including dropout and normalization
- Create a flexible architecture supporting variable sequence lengths
- Ensure GPU compatibility and efficient memory usage
- Achieve target performance metrics (AUC-ROC > 0.88)

### File Location
- **Primary File**: `src/models/lstm_model.py`
- **Dependencies**: `torch>=1.13.0`, `numpy>=1.23.0`

#### Core LSTM Features
- **Bidirectional LSTM layers** with forward and backward temporal processing
- **Configurable hidden size**: 32, 64, 128, 256 options
- **Variable depth**: 1-4 LSTM layers with residual connections for deeper networks
- **Dropout regularization**: Applied between LSTM layers and in classifier head
- **Gradient clipping**: Prevent exploding gradients during training

#### Normalization Options
- **Batch normalization**: For stable training dynamics
- **Layer normalization**: Alternative normalization for sequence data
- **Input normalization**: Z-score standardization integration

#### Classifier Head Design
- **Fully connected layers**: Multi-layer classifier with configurable dimensions
- **Activation functions**: ReLU, Sigmoid, Tanh options
- **Output layer**: Single neuron with sigmoid activation for binary classification
- **Regularization**: Dropout and optional weight decay

### Model Configuration Schema
```yaml
lstm_model:
  architecture:
    input_size: 5  # Number of sensor channels
    hidden_size: 64  # Hidden state dimension
    num_layers: 2  # Number of LSTM layers
    bidirectional: true  # Enable bidirectional processing
    dropout: 0.2  # Dropout rate between layers
    
  classifier:
    hidden_dims: [32, 16]  # Classifier hidden layer dimensions
    activation: "relu"  # Activation function
    dropout: 0.3  # Classifier dropout rate
    
  normalization:
    batch_norm: true  # Enable batch normalization
    layer_norm: false  # Enable layer normalization
    input_norm: true  # Enable input normalization
    
  regularization:
    weight_decay: 1e-4  # L2 regularization
    gradient_clip: 1.0  # Gradient clipping threshold
```

### 1. Enhanced SteelDefectLSTM Class
```python
class SteelDefectLSTM(nn.Module):
    """Enhanced LSTM model for sequence-based defect prediction"""
    
    def __init__(self, config: Dict[str, Any]):
        # Architecture parameters
        # Bidirectional LSTM implementation
        # Configurable classifier head
        # Normalization layers
        # Regularization components
```

#### Key Methods to Implement:
- `__init__()`: Complete architecture initialization
- `forward()`: Forward pass with all features
- `init_hidden()`: Hidden state initialization
- `get_attention_weights()`: For model interpretability
- `freeze_layers()`: For transfer learning
- `get_layer_outputs()`: For feature extraction

#### Bidirectional LSTM Implementation
- Forward and backward temporal processing
- Concatenated or summed hidden states
- Proper handling of variable sequence lengths
- Efficient memory usage for large sequences

#### Dynamic Sequence Handling
- Support for sequences of length 100-500 time steps
- Automatic padding and masking for variable lengths
- Batch processing optimization
- Memory-efficient data loading

#### Regularization Techniques
- **Dropout**: Applied between LSTM layers and in classifier
- **Batch/Layer Normalization**: Stabilize training dynamics
- **Weight Decay**: L2 regularization on all parameters
- **Gradient Clipping**: Prevent exploding gradients

#### Standard LSTM Configuration
```python
standard_config = {
    "input_size": 5,
    "hidden_size": 64,
    "num_layers": 2,
    "bidirectional": False,
    "dropout": 0.2
}
```

#### Bidirectional LSTM Configuration
```python
bidirectional_config = {
    "input_size": 5,
    "hidden_size": 64,
    "num_layers": 2,
    "bidirectional": True,
    "dropout": 0.2
}
```

#### Deep LSTM Configuration
```python
deep_config = {
    "input_size": 5,
    "hidden_size": 128,
    "num_layers": 4,
    "bidirectional": True,
    "dropout": 0.3,
    "residual_connections": True
}
```

### Model Performance Targets
- **AUC-ROC**: > 0.88 on test set
- **AUC-PR**: > 0.75 on test set
- **Training Time**: < 30 minutes on GPU
- **Inference Time**: < 200ms per sequence
- **Memory Usage**: < 4GB GPU memory during training

### Computational Efficiency
- **Batch Processing**: Support batch sizes 16-128
- **GPU Acceleration**: CUDA compatibility required
- **Memory Optimization**: Efficient gradient computation
- **Model Size**: < 50MB for deployment

### Unit Tests (`tests/test_lstm_model.py`)
```python
class TestSteelDefectLSTM:
    def test_model_initialization()
    def test_forward_pass()
    def test_bidirectional_processing()
    def test_variable_sequence_lengths()
    def test_gradient_computation()
    def test_hidden_state_initialization()
    def test_normalization_layers()
    def test_dropout_functionality()
    def test_gpu_compatibility()
    def test_model_serialization()
```

### Integration Tests
- Model training with synthetic data
- Gradient flow verification
- Memory usage profiling
- Performance benchmarking
- Cross-validation testing

### Phase 1: Core Architecture 
1. Enhance existing `SteelDefectLSTM` class
2. Implement bidirectional LSTM layers
3. Add configurable classifier head
4. Implement basic regularization

### Phase 2: Advanced Features 
1. Add normalization options
2. Implement dynamic sequence handling
3. Add gradient clipping and weight decay
4. Create model configuration system

### Phase 3: Optimization & Testing
1. GPU optimization and memory efficiency
2. Comprehensive unit testing
3. Performance benchmarking
4. Documentation and examples

### Code Documentation
- Comprehensive docstrings for all methods
- Type hints for all parameters
- Usage examples in docstrings
- Architecture diagram in module docstring

### Usage Examples
```python
# Basic usage
config = load_lstm_config("configs/lstm_config.yaml")
model = SteelDefectLSTM(config)

# Training setup
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
criterion = nn.BCEWithLogitsLoss()

# Forward pass
outputs = model(sequences)
loss = criterion(outputs, labels)
```

## Success Criteria
- [ ] Model successfully handles bidirectional LSTM processing
- [ ] Configurable architecture parameters work correctly
- [ ] All normalization options function properly
- [ ] Dropout and regularization prevent overfitting
- [ ] Variable sequence lengths are handled efficiently
- [ ] GPU training achieves target performance metrics
- [ ] Model size and inference time meet requirements
- [ ] All unit tests pass with > 90% code coverage
- [ ] Integration with existing training pipeline works
- [ ] Model serialization/loading functions correctly

## Dependencies & Environment
```python
torch>=1.13.0
torchvision>=0.14.0
numpy>=1.23.0
pyyaml>=6.0
tqdm>=4.64.0
pytest>=7.2.0
```

## Related Tasks
- Links to parent issue: Phase 3: Deep Learning Model Development (LSTM) #6
- Depends on: Sequence Feature Extraction Pipeline (Task 3.2)
- Enables: LSTM Training Pipeline (Task 3.4)
- Enables: Model Training Script (Task 3.5)

## Risk Mitigation
- **Memory Issues**: Implement gradient checkpointing for large models
- **Training Instability**: Use gradient clipping and learning rate scheduling
- **Overfitting**: Comprehensive regularization and early stopping
- **Performance**: Benchmark against baseline model requirements
- **Compatibility**: Ensure PyTorch version compatibility across environments
